Universal algebra first commit

shiranaiyo opened the pull request:
First commit on universal algebra. Credits to @plt-amy for coming up on how to do the `is-algebra` thing.

I'm not convinced on the names for the stuff under the sections `Abstract equations` and `Theories`, but otherwise I think it's fine.
I think that doing
```
group-Algebra-Group (((A , is-set-A) , models-A) , satisfies-A) =
```
is not common in this library but it really helps to reduce a lot of the boilerplate. Not sure if creating types equivalent to the standard finite types is fine. It certainly helps readability in my opinion.

shiranaiyo commented on src/universal-algebra/algebras.lagda.md:
```diff
@@ -0,0 +1,384 @@
+          ( eq-is-prop
+            ( is-prop-is-algebra
+              ( group-signature) {lzero} {l2} {group-Theory}
+              ( model-Algebra group-signature {lzero} {l2} {group-Theory} Alg))))))
```
I couldn't omit these types here despite them being "obvious" since `Alg` is an algebra on the a `group-Theory`, wondering if anyone has any idea on how to omit them.

EgbertRijke commented on src/universal-algebra/algebras.lagda.md:
```diff
@@ -0,0 +1,384 @@
+
+## Definitions
+
+### Signatures
```
I think there can be a separate file for signatures

EgbertRijke commented on src/universal-algebra/algebras.lagda.md:
````diff
@@ -0,0 +1,384 @@
+arity-operations-signature Sig = pr2 Sig
+```
+
+### Models
````
Separate file `models-signatures`

EgbertRijke commented on src/universal-algebra/algebras.lagda.md:
````diff
@@ -0,0 +1,384 @@
+  is-set-type-Model M = pr2 (set-Model M)
+```
+
+### Terms
````
seperate file `terms-signatures`

EgbertRijke commented on src/universal-algebra/algebras.lagda.md:
````diff
@@ -0,0 +1,384 @@
+  eval-vec m assign (x ∷ v) = eval-term m assign x ∷ eval-vec m assign v
+```
+
+### Abstract equations
````
separate file `abstract-equations`

EgbertRijke commented on src/universal-algebra/algebras.lagda.md:
````diff
@@ -0,0 +1,384 @@
+  rhs-Abstract-Equation = pr2
+```
+
+### Theories
````
separate file `algebraic-theories`

EgbertRijke commented on src/universal-algebra/algebras.lagda.md:
````diff
@@ -0,0 +1,384 @@
+  index-Abstract-Equation-Theory Th e = pr2 Th e
+```
+
+### Algebra
````
separate file `algebras-theories`

EgbertRijke commented on src/universal-algebra/algebras.lagda.md:
```diff
@@ -0,0 +1,384 @@
+
+## Examples
+
+### The algebra of groups
```
separate file `algebraic-theory-of-groups`

EgbertRijke commented:
Very nice PR!

EgbertRijke commented:
I hope you're safe over there. I heard there was a big earthquake in Ecuador/Peru.

fredrik-bakke commented on src/universal-algebra/algebras.lagda.md:
```diff
@@ -0,0 +1,384 @@
+# Cumulative hierarchy
```
Looks like you forgot to change the title as well.

shiranaiyo replied:
ooooops hahaha

EgbertRijke commented:
> First commit on universal algebra. Credits to @plt-amy for coming up on how to do the `is-algebra` thing.
> 
> I'm not convinced on the names for the stuff under the sections `Abstract equations` and `Theories`, but otherwise I think it's fine. I think that doing
> 
> ```
> group-Algebra-Group (((A , is-set-A) , models-A) , satisfies-A) =
> ```
> 
> is not common in this library but it really helps to reduce a lot of the boilerplate. Not sure if creating types equivalent to the standard finite types is fine. It certainly helps readability in my opinion.

I disagree that I it will reduce the boilerplate code. The way it is set up now is essentially using extensions of theories, which could also be a part of the universal algebra code. The theory of monoids is an extension of the theory of semigroups. This makes a model of the theory of monoids a semigroup equipped with an interpretation of the extended theories. 

Set up in this way, you'll be able to use all that you know about semigroups in the development of the theory of monoids. In your proposal, this possibility is discarded. So my suggestion would be to also define extensions of theories, models of extended theories, and so on.

shiranaiyo commented:
> I hope you're safe over there. I heard there was a big earthquake in Ecuador/Peru.

Yes, luckily it's far from where I am.

shiranaiyo commented:
Thanks all for your comments! I've refactored into multiple files (and changed the titles haha). The group theory in abstract algebra I'm leaving as it is for now, as I haven't formalized what extending an algebra means yet.

I'll be focusing on quotient algebras and the first homomorphism theorem now

fredrik-bakke commented on src/universal-algebra/signatures.lagda.md:
```diff
@@ -0,0 +1,82 @@
+
+## Idea
+
+A signature is a collection of function symbols with a given arity.
```
with given arities. (They can have different arities, correct?)

fredrik-bakke commented on src/universal-algebra/terms-signatures.lagda.md:
```diff
@@ -0,0 +1,104 @@
+# Terms
```
Terms over signatures

fredrik-bakke commented on src/universal-algebra/terms-signatures.lagda.md:
````diff
@@ -0,0 +1,104 @@
+# Terms
+
+```agda
+module universal-algebra.terms-signatures where
````
`universal-algebra.terms-over-signatures`

fredrik-bakke commented on src/universal-algebra/abstract-equations.lagda.md:
```diff
@@ -0,0 +1,39 @@
+# Abstract equations
```
Abstract equations over signatures

fredrik-bakke commented on src/universal-algebra/abstract-equations.lagda.md:
````diff
@@ -0,0 +1,39 @@
+# Abstract equations
+
+```agda
+module universal-algebra.abstract-equations where
````
`universal-algebra.abstract-equations-over-signatures`

fredrik-bakke reviewed (changes requested):
I know I am contradicting Egbert with some of these comments, but we usually include prepositions in our module names. E.g. `terms-over-signatures` as opposed to `terms-signatures`.

fredrik-bakke commented on src/universal-algebra/models-signatures.lagda.md:
````diff
@@ -0,0 +1,62 @@
+# Models
+
+```agda
+module universal-algebra.models-signatures where
````
or `universal-algebra.models-over-signatures`

fredrik-bakke commented:
A very cool PR by the way! I had to learn some of these things myself when I was giving a lecture on classifying topoi last semester :)

shiranaiyo commented on src/foundation/set-quotients.lagda.md:
```diff
@@ -211,6 +211,14 @@ module _
       ( reflecting-map-quotient-map R)
       ( is-surjective-and-effective-quotient-map R)
+
+  inv-precomp-set-quotient :
```
I tried to do it that way but failed. My current idea is doing it through the equivalences $$A^{n+1} \to A \cong A \to (A^n \to A) \cong A \to (\bar A^n \to \bar A) \cong \bar A \to (\bar A^n \to \bar A)$$
since the initial map respects the equivalence relation. That is, by induction together with the universal property of set quotients.

EgbertRijke replied:
What about defining n-ary operations on A recursively:

0-ary-op(A) := A
(n+1)-ary-op(A) := A -> n-ary-op(A)

shiranaiyo replied:
I did consider it but since everything is in terms of vectors (as that's the only way to do abstract equations), I'd have to go back and forth from vectors to n-ary-op constantly.

EgbertRijke commented on src/universal-algebra/homomorphisms-algebras.lagda.md:
```diff
@@ -0,0 +1,65 @@
+  ( Alg2 : Algebra Sig Th l4)
+  where
+
+  preserves-ops-Algebra :
```
`preserves-operations-Algebra`

A library that is riddled with abbreviations quickly becomes unreadable, so we try to avoid that.

EgbertRijke commented on src/universal-algebra/models-signatures.lagda.md:
```diff
@@ -0,0 +1,62 @@
+  Model : (l2 : Level) → UU (l1 ⊔ lsuc l2)
+  Model l2 = Σ (Set l2) (λ X → is-model-Set X)
+
+  set-Model : {l2 : Level} → Model l2 → Set l2
```
`set-Model-Signature`

EgbertRijke commented on src/universal-algebra/models-signatures.lagda.md:
```diff
@@ -0,0 +1,62 @@
+    is-model-Set (set-Model M)
+  is-model-set-Model M = pr2 M
+
+  type-Model : {l2 : Level} → Model l2 → UU l2
```
`type-Model-Signature`

EgbertRijke commented on src/universal-algebra/models-signatures.lagda.md:
```diff
@@ -0,0 +1,62 @@
+  type-Model : {l2 : Level} → Model l2 → UU l2
+  type-Model M = pr1 (set-Model M)
+
+  is-set-type-Model :
```
`is-set-type-Model-Signature`

shiranaiyo commented:
Thanks for you comment, I think I've covered all of them. The induced map $A^n \to A$ to $\bar A^n \to \bar A$ is turning out to be somewhat tricky, I'll keep at it for a few more hours.

fredrik-bakke commented on src/universal-algebra/homomorphisms-algebras.lagda.md:
````diff
@@ -0,0 +1,65 @@
+# Homomorphisms of algebras
+
+```agda
+module universal-algebra.homomorphisms-algebras where
````
`universal-algebra.homomorphisms-of-algebras`

Sorry, missed this one in my last review.

fredrik-bakke reviewed (approved):
Good work!
I see you missed some of the comments in my last review. I've resolved the ones you've already addressed, but please go over the ones that are still unresolved.

EgbertRijke commented:
When I defined the binary functoriality of set quotients, I also created a file about exponents of set quotients. It was very useful, because the binary functoriality followed directly from stuff in there. I expect that the exponents will be useful to you too.

fredrik-bakke commented:
Another remark. Personally, I would prefer variable names such as `S` or `signature` for a signature, `M` or `model` for a model over a signature, `T` or `theory` for a theory and so on. This is a matter taste, but `Sg` and `Th` throws me off when reading your code for some reason.

fredrik-bakke commented on src/universal-algebra/algebraic-theories.lagda.md:
```diff
@@ -10,25 +10,28 @@ module universal-algebra.algebraic-theories where
 ## Idea
 
-A theory is a family of abstract equations that should hold.
+A theory is a collection of abstract equations over a signature `S` that we
```
Note that you can also add the asterisks for emphasis if you want. This will be displayed like **algebraic theory**.

shiranaiyo commented:
> I see you missed some of the comments in my last review. I've resolved the ones you've already addressed, but please go over the ones that are still unresolved.

My bad, thanks for pointing it out, I think I've covered them now, as with the new ones.

> When I defined the binary functoriality of set quotients, I also created a file about exponents of set quotients. It was very useful, because the binary functoriality followed directly from stuff in there. I expect that the exponents will be useful to you too.

Thanks! I'll check it out.

> Another remark. Personally, I would prefer variable names such as S or signature for a signature, M or model for a model over a signature, T or theory for a theory and so on. This is a matter taste, but Sg and Th throws me off when reading your code for some reason.

I think both are used in the literature, you may be used to one that uses S, M and T? I'm following Crole's Categories for Types here.

fredrik-bakke commented:
My references were the elephant, and _Theories, Sites, Toposes_ by O. Caramello. How things are used in the literature and how we do them in the library is not the same thing though. Although of course, one informs the other.

fredrik-bakke commented:
> This PR should probably be wrapped up before you start to add quotients to it. It is already quite nice! If you can resolve the review comments, then we can merge it. On the other hand, if you start working on quotients in this same PR, then it might take a long while before it gets to a state where we can merge it.

I can add to this that if you want to start a new addition while a PR is still under review, you can create a new branch from your current PR branch and start working in that one so that you can build on the current work while not changing the current PR. When you make changes in the PR branch, you can simply merge those into your other branch as well.

fredrik-bakke commented:
I think this PR is ready to be merged now. Thank you for another excellent contribution!

shiranaiyo commented:
Thanks for your reviews! I really should read some topos theory already.

fredrik-bakke commented:
The concepts as you've defined them don't all directly translate to her setting though. For instance, she only considers models in categories, and calls models of a signature $\Sigma$ for $\Sigma$-structures, while a _model_ in her words is a model of a theory in a category.

fredrik-bakke commented:
Also, her definition of a signature is that of _multi-sorted signatures with relations_. In case this is not something that is considered in your reference.

shiranaiyo commented:
Thnaks! I'll check it out (eventually). Yeah, I've noticed that there's a lot of different ways to study "theories", all of them at different levels and with no obvious way to relate them neatly. Hopefully I'll try that at a later time, once I have a better grasp of this.
